### DISCUSSION TITLE
Unable to rebind WebSocketConnectionSource

### DISCUSSION THREAD

**Discussion by BrightKn1ght** [eclipse-theia/theia#13377] [Q&A] [answered]
We derived our own IDE from the theia project version 1.34.4. As one of our adaptions we added an authentication token between the browser and the theia backend based on bearer tokens.

We extended from WebSocketConnectionProvider  and have overriden the function createWebSocket to inject an auth token:
![our_adaption](https://github.com/eclipse-theia/theia/assets/154238723/11d3c608-18b2-415c-baec-ece94e680971)

 
We created our own theia plugin called theia-tiger which was able to rebind the existing WebSocketConnectionProvider with our adapted class AuthConnectionProvider before the class was constructed. At the start() method just our AuthConnectionProvider was loaded:
![theia1 34 4](https://github.com/eclipse-theia/theia/assets/154238723/0473c858-b0ae-4a4e-a68e-a92dd85497dd)

After PR https://github.com/eclipse-theia/theia/pull/13082/ the WebSocketConnectionProvider was replaced with WebSocketConnectionSource and also its instantiation time is now different:
![theia1 46 1](https://github.com/eclipse-theia/theia/assets/154238723/cd0b0375-1fff-4712-a248-e5d166d84620)

Our plugin theia-tiger  loads later and is not able to rebind the already created socket: 
![too_late](https://github.com/eclipse-theia/theia/assets/154238723/a5331ca2-aee6-4c0b-88df-74b36690d270)

I'm not an inversify expert. I would have the follwing questions:

1. Would it be possible with some inversify magic to still rebind the WebSocketConnectionProvider in our theia-tiger plugin?
2. If not how would you suggest to add an auth token to the browsers socket in the current release 1.46.1?

Thanks for your help!

### Additional Information

 

- Operating System:

- Theia Version: 1.46.1

**Comment by msujew** [comment, marked as answer]
Hey @BrightKn1ght,

you need to rebind the `WebSocketConnectionSource` inside of a custom `frontendPreload` module. As you have already noticed, binding it in a `frontend` module is too late.

**Reply by BrightKn1ght** [reply]
Hey @msujew,
thanks for your quick reply and the nice hint. I was not aware that this is possible since this `index.js` is a generated file. Is there an example available how to add a custom plugin to `frontendPreload` module?

**Reply by msujew** [reply]
Yes, see https://github.com/eclipse-theia/theia/blob/4b8ec2de6576702d1470d520e8ab195149046508/packages/core/package.json#L134

Note that `index.js` is generated from the modules found in your `package.json`, at least in parts.

**Reply by BrightKn1ght** [reply]
Nice, Thank you. I will give it a try if I can get it working.

**Reply by BrightKn1ght** [reply]
Thank you very much Mark! I could get it working by putting this piece of code to a custom `frontendPreload` module.

`export default new ContainerModule((bind, unbind, isBound, rebind) => {
    rebind(WebSocketConnectionSource).to(AuthWebSocketConnectionSource).inSingletonScope();
});`
